velox/functions/sparksql/Rand.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/functions/Macros.h"
+
+namespace facebook::velox::functions {
+
+template <typename T>
+struct RandFunction {
+  static constexpr bool is_deterministic = false;
+
+  std::optional<std::mt19937> generator;
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& /*config*/,
+      const int32_t* seed,
+      const int32_t* partitionIndex) {
+    if (seed == nullptr || partitionIndex == nullptr) {
+      VELOX_USER_FAIL(
+          "Spark rand/random function requires constant seed/partitionIndex!");
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& /*config*/,
+      const int64_t* seed,
+      const int32_t* partitionIndex) {
+    if (seed == nullptr || partitionIndex == nullptr) {
+      VELOX_USER_FAIL(
+          "Spark rand/random function requires constant seed/partitionIndex!");
+    }
+  }
+
+  // Partition index is NOT needed to differentiate generator as ThreadLocalPRNG
+  // is used.
+  FOLLY_ALWAYS_INLINE void call(double& result) {
+    result = folly::Random::randDouble01();
+  }
+
+  FOLLY_ALWAYS_INLINE void callNullable(
```

Comment:
Hi @mbasmanova, I was just aware of this comment.

Did you mean in `initialize` method (see copied below) `seed` is nullptr when `seed` is not constant or `seed` is a NULL value? If so, it indeed has issue for NULL value, as I expected `callNullable` should tackle NULL value case.

```
 FOLLY_ALWAYS_INLINE void initialize(
      const core::QueryConfig& /*config*/,
      const int32_t* seed,
      const int32_t* partitionIndex) {
    VELOX_USER_CHECK_NOT_NULL(
        seed, "Spark rand/random function requires constant seed.");
```

Reply:
I prefer to remove `initialize` method which is used to block non-constant input, as I found Spark framework doesn't allow non-constant input. Thus, with the current logic in `initialize` removed, we can allow NULL seed value to be handled by `callNullable`.

Reply:
CC: @laithsakka Laith, it would be great to enhance the framework to allow to distinguish between non-constant and null-constant arguments in initialize.

@PHILO-HE Would you file a GitHub to explain this limitation of the framework.

Reply:
Yes, just filed one issue: https://github.com/facebookincubator/velox/issues/6569.

Reply:
Thanks.